frame/contracts/src/wasm/runtime.rs

```diff
@@ -33,31 +35,67 @@ use sp_io::hashing::{
 	sha2_256,
 };
 
-/// The value returned from ext_call and ext_instantiate contract external functions if the call or
-/// instantiation traps. This value is chosen as if the execution does not trap, the return value
-/// will always be an 8-bit integer, so 0x0100 is the smallest value that could not be returned.
-const TRAP_RETURN_CODE: u32 = 0x0100;
+/// Every error that can be returned from a runtime API call.
+#[repr(u32)]
+pub enum ReturnCode {
+	/// API call successful.
+	Success = 0,
+	/// The called function trapped and has its state changes reverted.
+	/// In this case no output buffer is returned.
+	/// Can only be returned from `ext_call` and `ext_instantiate`.
+	CalleeTrapped = 1,
+	/// The called function ran to completion but decided to revert its state.
+	/// An output buffer is returned when one was supplied.
+	/// Can only be returned from `ext_call` and `ext_instantiate`.
+	CalleeReverted = 2,
+	/// The passed key does not exist in storage.
+	KeyNotFound = 3,
+}
+
+impl ConvertibleToWasm for ReturnCode {
+	type NativeType = Self;
+	const VALUE_TYPE: ValueType = ValueType::I32;
+	fn to_typed_value(self) -> sp_sandbox::Value {
+		sp_sandbox::Value::I32(self as i32)
+	}
+	fn from_typed_value(_: sp_sandbox::Value) -> Option<Self> {
+		// We will only ever send these values to wasm but never receive them.
+		unimplemented!()
+	}
+}
+
+impl From<ExecReturnValue> for ReturnCode {
+	fn from(from: ExecReturnValue) -> ReturnCode {
+		if from.flags.contains(ReturnFlags::REVERT) {
+			Self::CalleeReverted
+		} else {
+			Self::Success
+		}
+	}
+}
 
 /// Enumerates all possible *special* trap conditions.
 ///
 /// In this runtime traps used not only for signaling about errors but also
 /// to just terminate quickly in some cases.
 enum SpecialTrap {
 	/// Signals that trap was generated in response to call `ext_return` host function.
-	Return(Vec<u8>),
+	Return(u32, Vec<u8>),
 	/// Signals that trap was generated because the contract exhausted its gas limit.
 	OutOfGas,
 	/// Signals that a trap was generated in response to a succesful call to the
 	/// `ext_terminate` host function.
 	Termination,
 	/// Signals that a trap was generated because of a successful restoration.
 	Restoration,
+	/// Signals that a buffer supplied by the caller is too small to fit the output.
+	OutputBufferTooSmall,
```

Comment:
Nit: Adding this actually makes this enumeration to convey not only "special traps", since I see `OutputBufferTooSmall` to be just a plain error.

Reply:
Maybe we should rename that thing then? I am planning to use it even more in order to differantiate errors that originate in the supervisor from the ones that originates in the contract. We can even get them displayed in the UI using separate `decl_error!` entires for those.

Reply:
I renamed it to `TrapReason` and adjusted the enum and its comment to reflect properly that it is both for actual errors and "Special Traps".